File: LeetCode/34.Find_First_and_Last_Position_of_Element_in_Sorted_Array/34.Find_First_and_Last_Position_of_Element_in_Sorted_Array.swift
class Solution {
    func searchRange(_ nums: [Int], _ target: Int) -> [Int] {
        let n = nums.count
        var l = 0, r = n

        while l < r {
            let mid = (l + r) >> 1
            if nums[mid] < target {
                l = mid + 1
            } else {
                r = mid
            }
        }

        var res: [Int] = []
        if l < n, nums[l] == target {
            res.append(l)
        } else {
            return [-1, -1]
        }

        l = 0
        r = n
        while l < r {
            let mid = (l + r) >> 1
            if nums[mid] > target {
                r = mid
            } else {
                l = mid + 1
            }
        }

        res.append(l - 1)

        return res
    }
}

File: LeetCode/34.Find_First_and_Last_Position_of_Element_in_Sorted_Array/34.Find_First_and_Last_Position_of_Element_in_Sorted_Array.py
#!/usr/bin/env python3

import bisect


class Solution:
    def searchRange(self, nums: list[int], target: int) -> list[int]:
        n = len(nums)
        l, r = 0, n

        res = []
        while l < r:
            m = (l + r) // 2
            if nums[m] < target:
                l = m + 1
            else:
                r = m

        if l < n and nums[l] == target:
            res.append(l)
        else:
            return [-1, -1]

        l, r = 0, n
        while l < r:
            m = (l + r) // 2
            if target < nums[m]:
                r = m
            else:
                l = m + 1

        res.append(l - 1)
        return res


class Solution2:
    def searchRange(self, nums: list[int], target: int) -> list[int]:
        start = bisect.bisect_left(nums, target)
        if start == len(nums) or nums[start] != target:
            return [-1, -1]

        end = bisect.bisect_right(nums, target)
        return [start, end - 1]
